src/mast/mod.rs

```diff
@@ -0,0 +1,1109 @@
+
+                let res = if is_generic_parameter(&name.value) {
+                    let mtype = typed_fn_env.get_type_info(&name.value).unwrap();
+                    let mexpr = expr.to_mast(
+                        &mut self.ctx,
+                        &ExprKind::BigUInt(BigUint::from(mtype.value.unwrap())),
+                    );
+
+                    ExprMonoInfo::new(mexpr, Some(mtype.typ.clone()), mtype.value)
+                } else if is_type(&name.value) {
+                    let mtype = TyKind::Custom {
+                        module: module.clone(),
+                        name: name.value.clone(),
+                    };
+
+                    let mexpr = expr.to_mast(
+                        &mut self.ctx,
+                        &ExprKind::Variable {
+                            module: module.clone(),
+                            name: name.clone(),
+                        },
+                    );
+
+                    ExprMonoInfo::new(mexpr, Some(mtype), None)
+                } else if let Some(cst) = self.tast.const_info(&qualified) {
+                    // if it's a variable,
+                    // check if it's a constant first
+                    let bigint: BigUint = cst.value[0].into();
+                    let cst: u32 = bigint.clone().try_into().expect("biguint too large");
+                    let mexpr = expr.to_mast(&mut self.ctx, &ExprKind::BigUInt(bigint));
+
+                    ExprMonoInfo::new(mexpr, Some(TyKind::BigInt), Some(cst))
+                } else {
+                    // otherwise it's a local variable
+                    let mexpr = expr.to_mast(
+                        &mut self.ctx,
+                        &ExprKind::Variable {
+                            module: module.clone(),
+                            name: name.clone(),
+                        },
+                    );
+
+                    let mtype = typed_fn_env.get_type_info(&name.value).unwrap().clone();
+                    ExprMonoInfo::new(mexpr, Some(mtype.typ), mtype.value)
+                };
+
+                res
+            }
+
+            ExprKind::ArrayAccess { array, idx } => {
+                // get type of lhs
+                let array_mono = self.monomorphize_expr(array, typed_fn_env)?;
+                let id_mono = self.monomorphize_expr(idx, typed_fn_env)?;
+
+                // get type of element
+                let el_typ = match array_mono.typ.unwrap() {
+                    TyKind::Array(typkind, _) => Some(*typkind),
+                    _ => panic!("not an array"),
+                };
+
+                let mexpr = expr.to_mast(
+                    &mut self.ctx,
+                    &ExprKind::ArrayAccess {
+                        array: Box::new(array_mono.expr),
+                        idx: Box::new(id_mono.expr),
+                    },
+                );
+
+                // todo: check the bounds of the array
```

Comment:
ah, but only if monomorphization happened! We shouldn't repeat checks already done in the TAST

Reply:
actually circuit writer already does this check.

Reply:
let's write that in a comment there :o